### ISSUE TITLE
Dependency on Crypto

### ISSUE THREAD

**Issue by MyCADDev** [dylang/shortid#10]
Hi,
I want to use shortid on client side. Is it possible to remove dependency on crypto with the below code?

thanks
Harish

        function randomBytes(bytes) {
            var n = Math.pow(2, (bytes*8));
            return [Math.floor((Math.random() * n))];
        }

**Comment by jessegrosjean** [comment]
I'm interested in this also. I just need uniqueID's they don't need to be secure. I changed encode to look like this (replacing `randomBytes(1)1 with `Math.floor((Math.random() * 256))`) and in addition to no longer needing Crypto I'm also getting much better performance.

```
function encode(lookup, number) {
    var loopCounter = 0;
    var done;

    var str = '';

    while (!done) {
        str = str + lookup( ( (number >> (4 * loopCounter)) & 0x0f ) | (Math.floor((Math.random() * 256)) & 0x30)  );
        done = number < (Math.pow(16, loopCounter + 1 )  );
        loopCounter++;
    }
    return str;
}

module.exports = encode;
```

**Comment by anthonyettinger** [comment]
i'm curious why crypto has to do with generate a unique id.

**Comment by jessegrosjean** [comment]
@anthonyettinger I think it's because in some cases you don't want the IDs to be predictable because an attacker can exploit that in an attack. But in other cases (such as my use of shortID) you don't care, and in that case it would be really nice to have a non crypto option that's faster.

**Comment by flesler** [comment]
Not only faster, also sequential. Which is many databases [improves performance when retrieving](http://stackoverflow.com/questions/170346/what-are-the-performance-improvement-of-sequential-guid-over-standard-guid) and also enables queries like:
```sql
SELECT * FROM xxx WHERE id > 'Adsx232X'
```
For stuff like infinite scroll / paging.

**Comment by dylang** [comment]
I picked crypto because I was being *overly clever* getting random bytes to make the id difficult to predict. I think `Math.random` should be good enough and happy to switch to it to make it easier to support `browserify` for using `shortId` in the browser.

@EgoAleSum has expressed interest in making a PR to support `browserify` in #25. 

@flesler I'm not sure if this going to make any difference for performance. I can't think of a good reason to write a query like `id > 'Adsx232X'`, the comparison should be based on something `dateCreated > today-2days` which will be sequential.

**Comment by flesler** [comment]
The main point was the first, the gt comparison was just an extra :D

**Comment by EgoAleSum** [comment]
I've created the PR to make shortId able to run on the browser, see #28.
For the problem with crypto, I used this logic:
1. On Node.js, still use crypto
2. On modern browsers that support the Web Crypto API, use their random number generator which is "cryptographically strong".
3. Use Math.random() as a fallback.

**Comment by flesler** [comment]
Just to understand this a little more, what is the down-side of just using Math.random(), don't tell me the difference with crypto but rather how it affects the results on shortid More collisions? more predictable?

**Comment by EgoAleSum** [comment]
Math.random's results are not evenly distributed. Also, it suffers from many issues, like http://ifsec.blogspot.ca/2012/05/cross-domain-mathrandom-prediction.html

I didn't run tests to check exactly how that affects the uniqueness of shortId's, but for sure the more random the generated numbers are, the better.

At the end, however, all these fallbacks (crypto on Node, Web Crypto API and Math.random) took just around 20 lines of code to implement.

**Comment by EgoAleSum** [comment]
@flesler 
I have a bit more time now so I can provide a more detailed answer. Aside from security flaws as the one listed above, the statistical rationale is this.

shortid's are basically random numbers. They're represented as strings, but in essence they're random numbers.
The reason why they're unique is just because of probability. Indeed, in the default configuration a shortid is 7 characters long, each character being one of the 64 possible. That means that there are about 4.4*10e12 possible combinations (that is, 4 trillion, or million billion, possibilities).

Now, suppose for simplicity that a shortid had only 2 combinations: 0 and 1. With a "true" random number generator (RNG), each number would have the same probability (1/2 or 50%) to be picked each time.
That is: if you generated that number many times (say, a million) with a RNG, you would get 50% of the times zero and 50% of the times one (statistically).

Now, "true" RNG are extremely uncommon, and software normally uses "pseudo" random number generator (PRNG).
There are different ways to obtain a PRNG, and some are better than other.
Nodejs (crypto) and the Web Crypto API use "good" PRNG, which means that each number generated "always" comes out with the same probability (e.g. 50%/50% in the example above).
Math.random, instead, uses a worse PRNG, and certain numbers are more likely to appear than other (say, 49.5% times zero, 50.5% times one).

Because with Math.random some numbers are more likely to appear than others, the probability that a collision in shortid appears is more likely.
I don't have a degree in statistics (and I don't have a degree in computer science either, for what matters :) ) so I can't tell you "how much more likely" that is to happen, but in theory it is.

**Comment by hhubik** [comment]
Could be that I am doing something wrong, but I just got this when trying to use the library in my Angular2 (systemJS) app:
zone.js?1472702572155:101 GET http://localhost:5555/node_modules/crypto/package.json 404 (Not Found)

I did the following:
- got the latest shortid (2.2.6) via npm (I see it in my node_modules folder).
- in node_modules/shortid, ran browserify . > ./browser/shortid.js
- included the generated shortid.js (see the attached zip) in my Angular2 project.

[shortid.zip](https://github.com/dylang/shortid/files/448808/shortid.zip)

Any suggestions?
